libraries/aws/jobs/src/jobs.c

```diff
@@ -0,0 +1,558 @@
+ *
+ * @param[in] a  first character sequence
+ * @param[in] b  second character sequence
+ * @param[in] n  number of bytes
+ *
+ * @return JobsSuccess if the sequences are the same;
+ * JobsNoMatch otherwise
+ */
+static JobsStatus_t strnEq( const char * a,
+                            const char * b,
+                            size_t n )
+{
+    size_t i;
+
+    for( i = 0U; i < n; i++ )
+    {
+        if( a[ i ] != b[ i ] )
+        {
+            break;
+        }
+    }
+
+    return ( i == n ) ? JobsSuccess : JobsNoMatch;
+}
+
+/**
+ * @brief Wrap strnEq() with a check to compare two lengths.
+ */
+static JobsStatus_t strnnEq( const char * a,
+                             size_t aLength,
+                             const char * b,
+                             size_t bLength )
+{
+    JobsStatus_t ret = JobsNoMatch;
+
+    if( aLength == bLength )
+    {
+        ret = strnEq( a, b, aLength );
+    }
+
+    return ret;
+}
+
+/**
+ * @brief predicate returns true for a valid job ID character
+ */
+static bool_ isJobIdChar( char a )
+{
+    bool_ ret;
+
+    if( ( a == '-' ) || ( a == '_' ) )
+    {
+        ret = true;
+    }
+    else if( ( a >= '0' ) && ( a <= '9' ) )
+    {
+        ret = true;
+    }
+    else if( ( a >= 'A' ) && ( a <= 'Z' ) )
+    {
+        ret = true;
+    }
+    else if( ( a >= 'a' ) && ( a <= 'z' ) )
+    {
+        ret = true;
+    }
+    else
+    {
+        ret = false;
+    }
+
+    return ret;
+}
+
+/**
+ * @brief predicate returns true for a valid job ID string
+ */
+static bool_ isValidJobId( const char * jobId,
+                           uint16_t jobIdLength )
+{
+    bool_ ret = false;
+
+    if( ( jobId != NULL ) && ( jobIdLength > 0U ) &&
+        ( jobIdLength <= JOBS_JOBID_MAX_LENGTH ) )
+    {
+        size_t i;
+
+        for( i = 0; i < jobIdLength; i++ )
+        {
+            if( isJobIdChar( jobId[ i ] ) == false )
+            {
+                break;
+            }
+        }
+
+        ret = ( i == jobIdLength ) ? true : false;
+    }
+
+    return ret;
+}
+
+/**
+ * @brief Search for the API portion of a topic string in a table
+ *
+ * @param[in] topic  The topic string to check.
+ * @param[in] topicLength  The length of the topic string.
+ * @param[out] outApi  The jobs topic API value if present, e.g., #JobsUpdateSuccess.
+ * @param[out] outJobId  The beginning of the jobID in the topic string.
+ * @param[out] outJobIdLength  The length of the jobID in the topic string.
+ *
+ * @return #JobsSuccess if a matching topic was found;
+ * #JobsNoMatch if a matching topic was NOT found
+ *   (parameter outApi gets #JobsInvalidTopic ).
+ */
+static JobsStatus_t matchApi( const char * topic,
+                              size_t topicLength,
+                              JobsTopic_t * outApi,
+                              char ** outJobId,
+                              uint16_t * outJobIdLength )
+{
+    JobsStatus_t ret = JobsNoMatch;
+    char * p = ( char * ) topic;
```

Comment:
I don't think we need to cast away the `const`ness from `topic`. Can't both `p` and `jobId` be `const char *`s? Also, I think `p` could have a more descriptive name, e.g. `pMatchStart`, since it looks like `p` is the value we start matching from

Reply:
I moved the cast to the outJobId assignment.  I feel like char *p is as much an idiom as for(i=0;...) and with such a short function I prefer it.